pkg/aggregator/no_aggregation_stream_worker.go

```diff
@@ -0,0 +1,218 @@
+	if len(samples) == 0 {
+		return
+	}
+	w.samplesChan <- samples
+}
+
+func (w *noAggregationStreamWorker) stop(wait bool) {
+	var blockChan chan struct{}
+	if wait {
+		blockChan = make(chan struct{})
+	}
+
+	trigger := trigger{
+		time:      time.Now(),
+		blockChan: blockChan,
+	}
+
+	w.stopChan <- trigger
+
+	if wait {
+		<-blockChan
+	}
+}
+
+func (w *noAggregationStreamWorker) flush(wait bool) {
+	var blockChan chan struct{}
+	if wait {
+		blockChan = make(chan struct{})
+	}
+
+	trigger := trigger{
+		time:      time.Now(),
+		blockChan: blockChan,
+	}
+
+	w.flushChan <- trigger
+
+	if wait {
+		<-blockChan
+	}
+}
+
+// mainloop of the no aggregation stream worker:
+//   * it receives samples and counts how much it has sent to the serializer, if it has more than a given amount it stops
+//     streaming for the serializer to start sending the payloads to the forwarder, and then starts the streaming
+//     mainloop again
+//   * it also checks every 2 seconds if it has stopped receiving samples, if so, it stops streaming to the
+//     the serializer for a while in order to let the serializer sends the payloads up to the forwarder, and starts
+//     the streaming mainloop again
+//   * listens for a stop signal
+//   * listens for a flush signal
+// This is not ideal since the serializer should automatically takes the decision when to flush payloads to
+// the serializer but that's not how it works today, see noAggregationStreamWorker comment.
+func (w *noAggregationStreamWorker) run() {
+	log.Debugf("Starting streaming routine for the no-aggregation pipeline")
+
+	ticker := time.NewTicker(time.Second * 2)
+	defer ticker.Stop()
+	// 24h has no significance here, it's just a date "far enough" to avoid testing too much
+	// if we need to automatically flush and to keep this test simple with no extra boolean.
+	lastStream := time.Now().Add(time.Hour * 24)
+
+	logPayloads := config.Datadog.GetBool("log_payloads")
+	w.seriesSink, w.sketchesSink = createIterableMetrics(w.flushConfig, w.serializer, logPayloads, false)
+
+	stopped := false
+	var flushBlockChan chan struct{}
+	var stopBlockChan chan struct{}
+
+	for !stopped {
+		start := time.Now()
+		serializedSamples := 0
+
+		metrics.Serialize(
+			w.seriesSink,
+			w.sketchesSink,
+			func(seriesSink metrics.SerieSink, sketchesSink metrics.SketchesSink) {
+			mainloop:
+				for {
+					select {
+					// stop signal
+					case trigger := <-w.stopChan:
+						stopped = true
+						stopBlockChan = trigger.blockChan
+						break mainloop
+
+					// ticker regularly producing a flush signal if necessary
+					case <-ticker.C:
+						n := time.Now()
+						if serializedSamples > 0 && lastStream.Before(n.Add(-time.Second*1)) {
+							log.Debug("noAggregationStreamWorker: triggering an automatic payloads flush to the forwarder (no traffic since 1s)")
+							go w.flush(true)
+							lastStream = n.Add(time.Hour * 24)
+						}
+
+					// flush signal
+					case trigger := <-w.flushChan:
```

Comment:
You may not need a flush channel here or I miss something.

This part of the code is the [producer](https://github.com/DataDog/datadog-agent/blob/7.38.0/pkg/metrics/iterable_metrics.go#L117). When the producer finishes, `sendIterableSeries(w.serializer, start, serieSource)` returns, `serieConsumer` returns and `Serialize` returns.

When the Agent needs to flush, calling `return` should be enough.

Reply:
You miss a couple things, the flush channel is needed in this routine for two reasons:
  * this is the pattern we always use to avoid race conditions, any routine can call `flush()` and this mainloop routine will do it when it is not doing something else (such as receiving samples for instance)
  * it is indeed returning with the "break mainloop", I prefer this to a return in this case since it could confuse the reader, we're not returning for the main `run()` call but only this `Serializer` embedded lambda (btw this pattern in the serializer is once again really complex to use).

Reply:
Discussed offline: could be removed since returning from `Serialize` is enough in every case.